File: packages/server/src/parsers/template_parser.py
#!/usr/bin/env python3
"""
PDF Template Parser
Parses PDFs using learned template mappings
"""

import sys
import json
import re
from datetime import datetime
from typing import List, Dict, Any, Optional


def parse_date(value: str, date_format: Optional[str] = None) -> Optional[str]:
    """Parse date string to YYYY-MM-DD format"""
    if not value or not value.strip():
        return None

    s = value.strip()

    # Format mapping
    format_map = {
        'DD/MM/YYYY': '%d/%m/%Y',
        'DD-MM-YYYY': '%d-%m-%Y',
        'DD/MM/YY': '%d/%m/%y',
        'DD-MM-YY': '%d-%m-%y',
        'YYYY-MM-DD': '%Y-%m-%d',
        'MM/DD/YYYY': '%m/%d/%Y',
        'DD-MMM-YYYY': '%d-%b-%Y',
        'DD-MMM-YY': '%d-%b-%y',
    }

    # Try specified format first
    if date_format and date_format in format_map:
        try:
            dt = datetime.strptime(s, format_map[date_format])
            return dt.strftime('%Y-%m-%d')
        except ValueError:
            pass

    # Try common formats
    formats = [
        '%d/%m/%Y', '%d-%m-%Y', '%d/%m/%y', '%d-%m-%y',
        '%Y-%m-%d', '%m/%d/%Y', '%d-%b-%Y', '%d-%b-%y',
        '%d/%m/%Y', '%d %b %Y', '%d %B %Y',
    ]

    for fmt in formats:
        try:
            dt = datetime.strptime(s, fmt)
            return dt.strftime('%Y-%m-%d')
        except ValueError:
            continue

    return None


def parse_amount(value: str) -> Optional[float]:
    """Parse amount string to float"""
    if value is None or value == '':
        return None

    s = str(value).strip()
    if not s:
        return None

    # Check for DR/CR suffix
    is_debit = bool(re.search(r'\s*(dr|DR|Dr)\s*$', s))
    is_credit = bool(re.search(r'\s*(cr|CR|Cr)\s*$', s))
    s = re.sub(r'\s*(dr|DR|Dr|cr|CR|Cr)\s*$', '', s)

    # Check for negative in parentheses
    is_negative = s.startswith('(') and s.endswith(')')
    if is_negative:
        s = s[1:-1]

    # Remove currency symbols and spaces
    s = re.sub(r'[₹$€£\s]', '', s)

    # Remove commas
    s = s.replace(',', '')

    # Handle minus sign
    has_minus = s.startswith('-')
    if has_minus:
        s = s[1:]

    try:
        num = float(s)
    except ValueError:
        return None

    # Apply sign
    if is_negative or has_minus or is_debit:
        return -abs(num)
    elif is_credit:
        return abs(num)

    return num


def get_column_value(row: List[Any], source: str) -> Any:
    """Get column value from row by source"""
    match = re.match(r'^col_(\d+)$', source)
    if match:
        index = int(match.group(1))
        if index < len(row):
            return row[index]
    return None


def parse_pdf_with_template(pdf_path: str, mappings: Dict[str, Any], password: Optional[str] = None) -> Dict[str, Any]:
    """
    Parse PDF using template mappings

    Args:
        pdf_path: Path to PDF file
        mappings: Field mappings from template
        password: Optional password for encrypted PDF

    Returns:
        {
            "transactions": [...],
            "errors": [...],
            "rows_processed": int,
            "rows_skipped": int,
        }
    """
    try:
        import pdfplumber
    except ImportError:
        return {"error": "pdfplumber not installed. Run: pip install pdfplumber"}

    try:
        pdf_options = {}
        if password:
            pdf_options['password'] = password

        with pdfplumber.open(pdf_path, **pdf_options) as pdf:
            all_tables = []

            for page in pdf.pages:
                tables = page.extract_tables()
                for table in tables:
                    if table and len(table) > 1:
                        all_tables.extend(table)

            if not all_tables:
                return {"error": "No tables found in PDF"}

            # Skip header row (assumed to be first row with common header keywords)
            header_keywords = ['date', 'amount', 'balance', 'narration', 'description', 'debit', 'credit']
            start_row = 0
            for i, row in enumerate(all_tables[:5]):
                if row:
                    row_text = ' '.join(str(cell).lower() for cell in row if cell)
                    if any(kw in row_text for kw in header_keywords):
                        start_row = i + 1
                        break

            transactions = []
            errors = []
            rows_skipped = 0

            for row_idx, row in enumerate(all_tables[start_row:], start=start_row + 1):
                # Skip empty rows
                if not row or not any(cell and str(cell).strip() for cell in row):
                    rows_skipped += 1
                    continue

                try:
                    txn = {'raw_data': {}}

                    for field, mapping in mappings.items():
                        source = mapping.get('source', '')
                        fmt = mapping.get('format')
                        value = get_column_value(row, source)

                        # Store raw value
                        txn['raw_data'][field] = str(value) if value else None

                        if field == 'date':
                            parsed = parse_date(str(value) if value else '', fmt)
                            if parsed:
                                txn['date'] = parsed

                        elif field == 'valueDate':
                            parsed = parse_date(str(value) if value else '', fmt)
                            if parsed:
                                txn['valueDate'] = parsed

                        elif field == 'narration':
                            txn['narration'] = str(value).strip() if value else ''

                        elif field == 'reference':
                            if value:
                                txn['reference'] = str(value).strip()

                        elif field == 'withdrawal':
                            amount = parse_amount(str(value) if value else '')
                            if amount is not None and amount != 0:
                                txn['withdrawal'] = abs(amount)

                        elif field == 'deposit':
                            amount = parse_amount(str(value) if value else '')
                            if amount is not None and amount != 0:
                                txn['deposit'] = abs(amount)

                        elif field == 'amount':
                            amount = parse_amount(str(value) if value else '')
                            if amount is not None:
                                txn['amount'] = amount

                        elif field == 'balance':
                            amount = parse_amount(str(value) if value else '')
                            if amount is not None:
                                txn['balance'] = amount

                        elif field == 'transactionType':
                            if value:
                                txn['transactionType'] = str(value).strip()

                        elif field == 'category':
                            if value:
                                txn['category'] = str(value).strip()

                        elif field == 'merchant':
                            if value:
                                txn['merchant'] = str(value).strip()

                        elif field == 'cardNumber':
                            if value:
                                txn['cardNumber'] = str(value).strip()

                    # Validate required fields
                    if not txn.get('date'):
                        errors.append(f"Row {row_idx}: Missing or invalid date")
                        rows_skipped += 1
                        continue

                    if not txn.get('narration') and not txn.get('merchant'):
                        errors.append(f"Row {row_idx}: Missing narration/description")
                        rows_skipped += 1
                        continue

                    if (txn.get('withdrawal') is None and
                        txn.get('deposit') is None and
                        txn.get('amount') is None):
                        errors.append(f"Row {row_idx}: Missing amount")
                        rows_skipped += 1
                        continue

                    transactions.append(txn)

                except Exception as e:
                    errors.append(f"Row {row_idx}: {str(e)}")
                    rows_skipped += 1

            return {
                "transactions": transactions,
                "errors": errors[:50],  # Limit errors
                "rows_processed": len(all_tables) - start_row,
                "rows_skipped": rows_skipped,
            }

    except Exception as e:
        error_msg = str(e).lower()
        if 'password' in error_msg or 'encrypt' in error_msg:
            return {"error": "PDF is password protected. Please provide the correct password."}
        return {"error": str(e)}


def main():
    if len(sys.argv) < 3:
        print(json.dumps({"error": "Usage: template_parser.py <pdf_path> <mappings_json> [password]"}))
        sys.exit(1)

    pdf_path = sys.argv[1]
    mappings = json.loads(sys.argv[2])
    password = sys.argv[3] if len(sys.argv) > 3 else None

    result = parse_pdf_with_template(pdf_path, mappings, password)
    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: packages/server/src/parsers/template-parser.ts
/**
 * Template Parser
 * Parses files using learned template mappings
 */

import * as XLSX from 'xlsx';
import { execSync } from 'child_process';
import * as path from 'path';
import * as fs from 'fs';
import * as os from 'os';
import { fileURLToPath } from 'url';
import { LearnedTemplate } from '../db/schema/templates.js';
import dayjs from 'dayjs';
import customParseFormat from 'dayjs/plugin/customParseFormat.js';

dayjs.extend(customParseFormat);

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export interface ParsedTransaction {
  date: string;
  valueDate?: string;
  narration: string;
  reference?: string;
  withdrawal?: number;
  deposit?: number;
  amount?: number;
  balance?: number;
  transactionType?: string;
  category?: string;
  merchant?: string;
  cardNumber?: string;
  rawData?: Record<string, any>;
}

export interface TemplateParseResult {
  success: boolean;
  transactions: ParsedTransaction[];
  errors: string[];
  rowsProcessed: number;
  rowsSkipped: number;
}

/**
 * Parse date string according to format
 */
function parseDate(value: any, format?: string): string | null {
  if (!value) return null;

  const str = String(value).trim();
  if (!str) return null;

  // Try specified format first
  if (format && format !== 'auto') {
    const formatMap: Record<string, string> = {
      'DD/MM/YYYY': 'DD/MM/YYYY',
      'DD-MM-YYYY': 'DD-MM-YYYY',
      'DD/MM/YY': 'DD/MM/YY',
      'DD-MM-YY': 'DD-MM-YY',
      'YYYY-MM-DD': 'YYYY-MM-DD',
      'MM/DD/YYYY': 'MM/DD/YYYY',
      'DD-MMM-YYYY': 'DD-MMM-YYYY',
      'DD-MMM-YY': 'DD-MMM-YY',
    };

    const dayjsFormat = formatMap[format];
    if (dayjsFormat) {
      const parsed = dayjs(str, dayjsFormat);
      if (parsed.isValid()) {
        return parsed.format('YYYY-MM-DD');
      }
    }
  }

  // Try common formats
  const formats = [
    'DD/MM/YYYY', 'DD-MM-YYYY', 'DD/MM/YY', 'DD-MM-YY',
    'YYYY-MM-DD', 'MM/DD/YYYY', 'DD-MMM-YYYY', 'DD-MMM-YY',
    'D/M/YYYY', 'D-M-YYYY',
  ];

  for (const fmt of formats) {
    const parsed = dayjs(str, fmt, true);
    if (parsed.isValid()) {
      return parsed.format('YYYY-MM-DD');
    }
  }

  // Try native Date parsing as fallback
  const nativeDate = new Date(str);
  if (!isNaN(nativeDate.getTime())) {
    return dayjs(nativeDate).format('YYYY-MM-DD');
  }

  return null;
}

/**
 * Parse amount string to number
 */
function parseAmount(value: any): number | null {
  if (value === null || value === undefined || value === '') {
    return null;
  }

  // If already a number
  if (typeof value === 'number') {
    return value;
  }

  let str = String(value).trim();
  if (!str) return null;

  // Check for DR/CR suffix (Indian banking)
  const isDebit = /\s*(dr|DR|Dr)\s*$/.test(str);
  const isCredit = /\s*(cr|CR|Cr)\s*$/.test(str);
  str = str.replace(/\s*(dr|DR|Dr|cr|CR|Cr)\s*$/g, '');

  // Check for negative in parentheses: (1,234.56)
  const isNegative = str.startsWith('(') && str.endsWith(')');
  if (isNegative) {
    str = str.slice(1, -1);
  }

  // Remove currency symbols and spaces
  str = str.replace(/[₹$€£\s]/g, '');

  // Remove commas
  str = str.replace(/,/g, '');

  // Handle minus sign
  const hasMinusSign = str.startsWith('-');
  if (hasMinusSign) {
    str = str.slice(1);
  }

  const num = parseFloat(str);
  if (isNaN(num)) {
    return null;
  }

  // Apply sign
  let result = num;
  if (isNegative || hasMinusSign || isDebit) {
    result = -Math.abs(num);
  } else if (isCredit) {
    result = Math.abs(num);
  }

  return result;
}

/**
 * Get column value from row by source
 */
function getColumnValue(row: any[], source: string): any {
  // Source format: "col_0", "col_1", etc.
  const match = source.match(/^col_(\d+)$/);
  if (match) {
    const index = parseInt(match[1], 10);
    return row[index];
  }
  return null;
}

/**
 * Parse Excel file using template
 */
async function parseExcelWithTemplate(
  buffer: Buffer,
  template: LearnedTemplate
): Promise<TemplateParseResult> {
  const workbook = XLSX.read(buffer, { type: 'buffer' });
  const sheet = workbook.Sheets[workbook.SheetNames[0]];
  const data: any[][] = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: '' });

  const mappings = JSON.parse(template.fieldMappings);
  const sampleHeaders = template.sampleHeaders ? JSON.parse(template.sampleHeaders) : [];

  // Find header row (match with sample headers)
  let headerRowIndex = 0;
  if (sampleHeaders.length > 0) {
    for (let i = 0; i < Math.min(15, data.length); i++) {
      const row = data[i];
      const rowStr = row.map((c: any) => String(c || '').toLowerCase().trim());
      const headerStr = sampleHeaders.map((h: string) => h.toLowerCase().trim());

      // Check if this row matches sample headers
      const matches = headerStr.filter((h: string) => rowStr.includes(h)).length;
      if (matches >= Math.min(3, headerStr.length * 0.5)) {
        headerRowIndex = i;
        break;
      }
    }
  }

  const transactions: ParsedTransaction[] = [];
  const errors: string[] = [];
  let rowsSkipped = 0;

  // Process data rows
  for (let i = headerRowIndex + 1; i < data.length; i++) {
    const row = data[i];

    // Skip empty rows
    if (!row.some((cell: any) => cell !== null && cell !== undefined && String(cell).trim() !== '')) {
      rowsSkipped++;
      continue;
    }

    try {
      const txn: ParsedTransaction = {
        date: '',
        narration: '',
        rawData: {},
      };

      // Extract fields according to mappings
      for (const [field, mapping] of Object.entries(mappings)) {
        const { source, format } = mapping as { source: string; format?: string };
        const value = getColumnValue(row, source);

        // Store raw value
        txn.rawData![field] = value;

        switch (field) {
          case 'date':
            const parsedDate = parseDate(value, format);
            if (parsedDate) {
              txn.date = parsedDate;
            }
            break;

          case 'valueDate':
            const parsedValueDate = parseDate(value, format);
            if (parsedValueDate) {
              txn.valueDate = parsedValueDate;
            }
            break;

          case 'narration':
            txn.narration = String(value || '').trim();
            break;

          case 'reference':
            txn.reference = value ? String(value).trim() : undefined;
            break;

          case 'withdrawal':
            const withdrawal = parseAmount(value);
            if (withdrawal !== null && withdrawal !== 0) {
              txn.withdrawal = Math.abs(withdrawal);
            }
            break;

          case 'deposit':
            const deposit = parseAmount(value);
            if (deposit !== null && deposit !== 0) {
              txn.deposit = Math.abs(deposit);
            }
            break;

          case 'amount':
            const amount = parseAmount(value);
            if (amount !== null) {
              txn.amount = amount;
            }
            break;

          case 'balance':
            const balance = parseAmount(value);
            if (balance !== null) {
              txn.balance = balance;
            }
            break;

          case 'transactionType':
            txn.transactionType = value ? String(value).trim() : undefined;
            break;

          case 'category':
            txn.category = value ? String(value).trim() : undefined;
            break;

          case 'merchant':
            txn.merchant = value ? String(value).trim() : undefined;
            break;

          case 'cardNumber':
            txn.cardNumber = value ? String(value).trim() : undefined;
            break;
        }
      }

      // Validate required fields
      if (!txn.date) {
        errors.push(`Row ${i + 1}: Missing or invalid date`);
        rowsSkipped++;
        continue;
      }

      if (!txn.narration && !txn.merchant) {
        errors.push(`Row ${i + 1}: Missing narration/description`);
        rowsSkipped++;
        continue;
      }

      // Need at least one amount field
      if (txn.withdrawal === undefined && txn.deposit === undefined && txn.amount === undefined) {
        errors.push(`Row ${i + 1}: Missing amount`);
        rowsSkipped++;
        continue;
      }

      transactions.push(txn);
    } catch (err: any) {
      errors.push(`Row ${i + 1}: ${err.message}`);
      rowsSkipped++;
    }
  }

  return {
    success: transactions.length > 0,
    transactions,
    errors,
    rowsProcessed: data.length - headerRowIndex - 1,
    rowsSkipped,
  };
}

/**
 * Parse PDF file using template
 */
async function parsePDFWithTemplate(
  filePath: string,
  template: LearnedTemplate,
  password?: string
): Promise<TemplateParseResult> {
  const scriptPath = path.join(__dirname, 'template_parser.py');

  // Check if script exists
  if (!fs.existsSync(scriptPath)) {
    throw new Error('PDF parser script not found');
  }

  const mappings = JSON.parse(template.fieldMappings);

  try {
    const args = password
      ? `"${filePath}" '${JSON.stringify(mappings)}' "${password}"`
      : `"${filePath}" '${JSON.stringify(mappings)}'`;

    const result = execSync(`python3 "${scriptPath}" ${args}`, {
      encoding: 'utf-8',
      maxBuffer: 50 * 1024 * 1024,
      timeout: 120000,
    });

    const parsed = JSON.parse(result.trim());

    if (parsed.error) {
      throw new Error(parsed.error);
    }

    return {
      success: parsed.transactions.length > 0,
      transactions: parsed.transactions,
      errors: parsed.errors || [],
      rowsProcessed: parsed.rows_processed || 0,
      rowsSkipped: parsed.rows_skipped || 0,
    };
  } catch (error: any) {
    throw new Error(`Failed to parse PDF: ${error.message}`);
  }
}

/**
 * Parse file using learned template
 */
export async function parseWithTemplate(
  buffer: Buffer,
  template: LearnedTemplate,
  filePath?: string,
  password?: string
): Promise<TemplateParseResult> {
  const fileType = template.fileType;

  if (fileType === 'pdf') {
    if (!filePath) {
      // Create temp file
      const tempPath = path.join(os.tmpdir(), `parse-${Date.now()}.pdf`);
      fs.writeFileSync(tempPath, buffer);
      try {
        return await parsePDFWithTemplate(tempPath, template, password);
      } finally {
        fs.unlinkSync(tempPath);
      }
    }
    return parsePDFWithTemplate(filePath, template, password);
  }

  if (fileType === 'xlsx' || fileType === 'xls' || fileType === 'csv') {
    return parseExcelWithTemplate(buffer, template);
  }

  throw new Error(`Unsupported file type: ${fileType}`);
}

/**
 * Check if a file matches a template's detection patterns
 */
export function matchesTemplate(
  content: string,
  filename: string,
  template: LearnedTemplate
): { matched: boolean; confidence: number } {
  const patterns = JSON.parse(template.detectionPatterns);
  let score = 0;

  // Check text patterns
  if (patterns.textPatterns && content) {
    const contentLower = content.toLowerCase();
    for (const pattern of patterns.textPatterns) {
      if (contentLower.includes(pattern.toLowerCase())) {
        score += 10;
      }
    }
  }

  // Check filename patterns
  if (patterns.filenamePatterns && filename) {
    const filenameLower = filename.toLowerCase();
    for (const pattern of patterns.filenamePatterns) {
      if (filenameLower.includes(pattern.toLowerCase())) {
        score += 5;
      }
      try {
        const regex = new RegExp(pattern, 'i');
        if (regex.test(filename)) {
          score += 8;
        }
      } catch {
        // Invalid regex, skip
      }
    }
  }

  return {
    matched: score >= 10,
    confidence: Math.min(score / 30, 1), // Normalize to 0-1
  };
}
